File: host/cnode/node.cpp
#include <cstring>
#include <string>
#include <iostream>
#include <sstream>
#include <boost/log/trivial.hpp>

#include "node.h"
#include "input.h"
#include "output.h"
#include "servo.h"
#include "utils.h"

Node::Node(int a, int i, int o, int s) {
	std::stringstream n;
	n << "node" << a;
	nodeInit(n.str(), a, i, o, s);
}

Node::Node(std::string nm, int a, int i, int o, int s) {
	nodeInit(nm, a, i, o, s);
}

void Node::nodeInit(std::string nm, int a, int i, int o, int s) {
	name = nm;
	addr = a;
	ninputs = i;
	nibits = i*NIBITS;
	noutputs = o;
	nobits = o*NOBITS;
	nservos = s;
	nsbits = s*16;

	errorCount = 0;

	BOOST_LOG_TRIVIAL(info) << __func__ << ": " << "Node " << name << " at address " << addr << " created with config i:" << ninputs << "(" << nibits
					<< ")/o:" << noutputs << "(" << nobits << ")/s:" << nservos << "(" << nsbits << ")";
	for (int i=0; i<nibits; i++) {
		inputs[i] = new Input();
		inputs[i]->currentState = true;
	}
	for (int i=0; i<nobits; i++) {
		outputs[i] = new Output();
		outputs[i]->currentState = false;
	}
	for (int i=0; i<nsbits; i++) {
		servos[i] = new Servo();
	}
}

void Node::setName(std::string nm) {
	name = nm;
}

int Node::getAddr(void) {
	return addr;
}

std::string Node::GetConfig(void) {
	std::ostringstream rpt;
	rpt << "{\"name\":\"" << name << "\",\"address\":" << addr << ",\"input\":" << nibits << ",\"output\":" << nobits
		<< ",\"servo\":" << nsbits << "}";
	return rpt.str();
}

int Node::setInputStates(bool *bvals, int *idx, int n) {
	int diffs;

	diffs = 0;
	for (int i=0; i<n; i++) {
		bool bv = bvals[i];
		int ix = idx[i];

		if (ix < 0 || ix >= nibits) {
			BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "input index of " << ix << " is out of range";
			return diffs;
		}
		int ov = inputs[ix]->currentState;
		if (ov != bv) {
			diffs++;
			inputs[ix]->currentState = bv;
		}
	}
	return diffs;
}

std::string Node::InputsReport(void) {
	std::ostringstream rpt;
	rpt << "{\"inputs\":{\"address\":" << addr << ",\"count\":" << nibits << ",\"values\":[";
	for (int i = 0; i<nibits; i++) {
		if (i != 0)
			rpt << ", ";
		rpt << (inputs[i]->currentState ? "true" : "false");
	}
	rpt << "]}}";
	return rpt.str();
}

std::string Node::InputsReportDelta(int *ix, int nix) {
	std::ostringstream rpt;
	rpt << "{\"inputs\":{\"address\":" << addr << ",\"count\":" << nix << ",\"delta\":true,\"values\":[";
	for (int i = 0; i<nix; i++) {
		if (i != 0)
			rpt << ", ";
		int idx = ix[i];
		rpt << "[" << idx << ", " << (inputs[idx]->currentState ? "true" : "false") << "]";
	}
	rpt << "]}}";
	return rpt.str();
}

int Node::setOutputStates(bool *vals, int n) {
	int diffs;

	diffs = 0;
	for (int i=0; i<n; i++) {
		bool bv = vals[i];
		bool ov = outputs[i]->currentState;
		if (ov != bv) {
			diffs++;
			outputs[i]->currentState = bv;
		}
	}
	return diffs;
}

std::string Node::OutputOn(int ox) {
	std::ostringstream rpt;
	if (ox < 0 || ox >= nobits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "output index of " << ox << " is out of range";
		return "";
	}
	bool ov = outputs[ox]->currentState;
	if (ov) 
		return ""; // already on
	outputs[ox]->currentState = true;
	rpt << "{\"outputs\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"values\":[";
	rpt << "[" << ox << ", true]]}}";
	return rpt.str();
}

std::string Node::OutputOff(int ox) {
	std::ostringstream rpt;
	if (ox < 0 || ox >= nobits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "output index of " << ox << " is out of range";
		return "";
	}
	bool ov = outputs[ox]->currentState;
	if (!ov) 
		return ""; // already off
	outputs[ox]->currentState = false;
	rpt << "{\"outputs\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"values\":[";
	rpt << "[" << ox << ", false]]}}";
	return rpt.str();
}

std::string Node::OutputsReport(void) {
	std::ostringstream rpt;
	rpt << "{\"outputs\":{\"address\":" << addr << ",\"count\":" << nibits << ",\"values\":[";
	for (int i = 0; i<nobits; i++) {
		if (i != 0)
			rpt << ", ";
		rpt << (outputs[i]->currentState ? "true" : "false");
	}
	rpt << "]}}";
	return rpt.str();
}

int Node::setServoValues(short *nrm, short *rev, short *ini, short *cur, int n) {
	int diffs;

	diffs = 0;
	for (int i=0; i<n; i++) {
		if ((servos[i]->normal != nrm[i]) || (servos[i]->reverse != rev[i]) || (servos[i]->initial != ini[i]) || (servos[i]->current != cur[i]))
			diffs++;
		servos[i]->normal = nrm[i];
		servos[i]->reverse = rev[i];
		servos[i]->initial = ini[i];
		servos[i]->current = cur[i];
	}
	return diffs;
}

std::string Node::ServosReport(void) {
	std::ostringstream rpt;
	rpt << "{\"servos\":{\"address\":" << addr << ",\"count\":" << nsbits << ",\"values\":[";
	for (int i = 0; i<nsbits; i++) {
		if (i != 0)
			rpt << ", ";
		rpt << "[ " << servos[i]->normal << ", " << servos[i]->reverse << ", " << servos[i]->initial << ", " << servos[i]->current << "]";
	}
	rpt << "]}}";
	return rpt.str();
}

std::string Node::TurnoutNormal(int tx) {
	std::ostringstream rpt;
	if (tx < 0 || tx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << tx << " is out of range";
		return "";
	}
	if (servos[tx]->current == servos[tx]->normal)
		return ""; // already normal

	servos[tx]->current = servos[tx]->normal;
	rpt << "{\"servos\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"values\":[";
	rpt << "[" << tx << ", " << servos[tx]->current << "]]}}";
	return rpt.str();
}

std::string Node::TurnoutReverse(int tx) {
	std::ostringstream rpt;
	if (tx < 0 || tx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << tx << " is out of range";
		return "";
	}
	if (servos[tx]->current == servos[tx]->reverse)
		return ""; // already reverse

	servos[tx]->current = servos[tx]->reverse;
	rpt << "{\"servos\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"values\":[";
	rpt << "[" << tx << ", " << servos[tx]->current << "]]}}";
	return rpt.str();
}

bool Node::isTurnoutReversed(int tx) {
	if (tx < 0 || tx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << tx << " is out of range";
		return false;
	}
	return servos[tx]->current == servos[tx]->reverse;
}

bool Node::isTurnoutNormal(int tx) {
	if (tx < 0 || tx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << tx << " is out of range";
		return false;
	}
	return servos[tx]->current == servos[tx]->normal;
}

std::string Node::SetTurnoutLimits(int tx, int n, int r, int ini) {
	std::ostringstream rpt;
	if (tx < 0 || tx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << tx << " is out of range";
		return "";
	}
	servos[tx]->normal = n;
	servos[tx]->reverse = r;
	servos[tx]->initial = ini;
	rpt << "{\"servos\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"limits\":true,\"values\":[";
	rpt << "[" << tx << ", " << n << ", " << r << ", " << ini << "]]}}";
	return rpt.str();
}

std::string Node::ServoAngle(int sx, int ang) {
	std::ostringstream rpt;
	if (sx < 0 || sx > nsbits) {
		BOOST_LOG_TRIVIAL(error) << __func__ << ": " << "servo index of " << sx << " is out of range";
		return "";
	}
	if (servos[sx]->current == ang)
		return ""; // already at angle

	servos[sx]->current = ang;
	rpt << "{\"servos\":{\"address\":" << addr << ",\"count\":1,\"delta\":true,\"values\":[";
	rpt << "[" << sx << ", " << servos[sx]->current << "]]}}";
	return rpt.str();
}

File: host/nodetrigger/src/node.py
#!/usr/bin/env python3

class Output:
	def __init__(self):
		self.value = False
		
	def setValue(self, nv):
		self.value = nv
		
	def getValue(self):
		print("output level, returning %s" % self.value)
		return self.value
		
class Input:
	def __init__(self):
		self.value = True
		
	def setValue(self, nv):
		self.value = nv
		
	def getValue(self):
		print("input level, returning %s" % self.value)
		return self.value
	
class Servo:
	def __init__(self):
		self.normal = 0
		self.reverse = 0
		self.initial = 0
		self.current = 0
		
	def setValues(self, n, r, i, c):
		self.normal = n
		self.reverse = r
		self.initial = i
		self.current = c
		
	def getValues(self):
		return self.normal, self.reverse, self.initial, self.current
	
	def setLimits(self, n, r, i):
		self.normal = n
		self.reverse = r
		if i is None:
			self.initial = self.normal
		else:
			self.initial = i
			
	def setCurrent(self, c):
		self.current = c
		
class Node:
	def __init__(self, addr, nm, i, o, s):
		self.addr = addr
		self.name = nm
		
		self.inputs = []
		for _ in range(i):
			self.inputs.append(Input())
		self.ninputs = i
		
		self.outputs = []
		for _ in range(o):
			self.outputs.append(Output())
		self.noutputs = o
		
		self.servos = []
		for _ in range(s):
			self.servos.append(Servo())
		self.nservos = s
		#print("add node %s a:%d i:%d o:%d s:%d" % (nm, addr, i, o, s))
		
	def getAddr(self):
		return self.addr
	
	def getName(self):
		return self.name
	
	def getOutput(self, n):
		if n < 0 or n > self.noutputs:
			print("outputs index of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return False
		
		print("node level: %d" % n)		
		return self.outputs[n].getValue()
	
	def setOutputs(self, n, vals):
		if n < 0 or n > self.noutputs:
			print("outputs count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return

		diffs = 0		
		for i in range(n):
			if self.outputs[i].getValue() != vals[i]:
				diffs += 1
				self.outputs[i].setValue(vals[i])
				
		return diffs
			
	def setOutputsDelta(self, n, vals):
		if n < 0 or n > self.noutputs:
			print("outputs count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		diffs = 0
		for i in range(n):
			if self.outputs[vals[i][0]].getValue() != vals[i][1]:
				diffs += 1
				self.outputs[vals[i][0]].setValue(vals[i][1])
				
		return diffs
			
	def getInput(self, n):
		if n < 0 or n > self.ninputs:
			print("inputs index of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return True

		print("node level: %d" % n)		
		return self.inputs[n].getValue()
			
	def setInputs(self, n, vals):
		if n < 0 or n > self.ninputs:
			print("inputs count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		diffs = 0
		for i in range(n):
			if self.inputs[i].getValue() != vals[i]:
				diffs += 1
				self.inputs[i].setValue(vals[i])
				
		return diffs
			
	def setInputsDelta(self, n, vals):
		if n < 0 or n > self.noutputs:
			print("inputs count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		diffs = 0
		for i in range(n):
			if self.inputs[vals[i][0]].getValue() != vals[i][1]:
				diffs += 1
				self.inputs[vals[i][0]].setValue(vals[i][1])
				
		return diffs
			
	def setServos(self, n, vals):
		if n < 0 or n > self.nservos:
			print("servos count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		for i in range(n):
			self.servos[i].setValues(vals[i][0], vals[i][1], vals[i][2], vals[i][3])
			
	def setServosDelta(self, n, vals):
		if n < 0 or n > self.nservos:
			print("servos count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		for i in range(n):
			self.servos[vals[i][0]].setCurrent(vals[i][1])
			
	def setServosLimits(self, n, vals):
		if n < 0 or n > self.nservos:
			print("servos count of %d is out of range for node %s(%d)" % (n, self.name, self.addr))
			return
		
		for i in range(n):
			self.servos[vals[i][0]].setLimits(vals[i][1], vals[i][2], vals[i][3])
			
